### PR TITLE
Switch to ES Map/Set internally

### PR DESCRIPTION
**Pull request by rbuckton**
Currently leaving this as a Draft PR for discussion. This improves our `Map` shim to support non-string keys, and adds a shim for `Set` so that we might consider using it in the compiler as well.

These shims exist for a vanishingly rare corner case of TypeScript running in an environment without native implementations (at this point, primarily just IE).

### CODE REVIEW

**Comment by rbuckton** [comment]
@typescript-bot perf test
@typescript-bot run dt
@typescript-bot test this
@typescript-bot user test this

File: src/compiler/binder.ts [hunk 1]
```diff
@@ -15,7 +15,7 @@ namespace ts {
         referenced: boolean;
     }
 
-    export function getModuleInstanceState(node: ModuleDeclaration, visited?: Map<ModuleInstanceState | undefined>): ModuleInstanceState {
+    export function getModuleInstanceState(node: ModuleDeclaration, visited?: Map<string, ModuleInstanceState | undefined>): ModuleInstanceState {
```
Comment (sheetalkamat) [inline src/compiler/binder.ts:18]: Cab you make these Map<Path, ModuleInstance> and so on instead. (For all builder maps

File: src/compiler/builder.ts [hunk 2]
```diff
@@ -24,11 +24,11 @@ namespace ts {
         /**
          * Cache of bind and check diagnostics for files with their Path being the key
          */
-        semanticDiagnosticsPerFile?: ReadonlyMap<readonly ReusableDiagnostic[] | readonly Diagnostic[]> | undefined;
+        semanticDiagnosticsPerFile?: ReadonlyMap<string, readonly ReusableDiagnostic[] | readonly Diagnostic[]> | undefined;
         /**
          * The map has key by source file's path that has been changed
          */
-        changedFilesSet?: ReadonlyMap<true>;
+        changedFilesSet?: ReadonlySet<string>;
```
Comment (sheetalkamat) [inline src/compiler/builder.ts:31]: Again `changedFilesSet<Path>`

File: src/compiler/core.ts [hunk 3]
```diff
@@ -1587,7 +1662,7 @@ namespace ts {
      * Case-sensitive comparisons compare both strings one code-point at a time using the integer
      * value of each code-point after applying `toUpperCase` to each string. We always map both
      * strings to their upper-case form as some unicode characters do not properly round-trip to
-     * lowercase (such as `ẞ` (German sharp capital s)).
```
Comment (sheetalkamat) [inline src/compiler/core.ts:1590]: what changed here?

File: src/compiler/corePublic.ts [hunk 4]
```diff
@@ -22,54 +22,65 @@ namespace ts {
...
-        // Internet Explorer's Map doesn't support iteration, so don't use it.
-        // Natives
-        // NOTE: TS doesn't strictly allow in-line declares, but if we suppress the error, the declaration
-        // is still used for typechecking _and_ correctly elided, which is out goal, as this prevents us from
-        // needing to pollute an outer scope with a declaration of `Map` just to satisfy the checks in this function
-        //@ts-ignore
-        declare const Map: (new <T>() => Map<T>) | undefined;
-        // eslint-disable-next-line no-in-operator
-        return typeof Map !== "undefined" && "entries" in Map.prototype ? Map : undefined;
+    export interface ReadonlySet<T> {
```
Comment (sheetalkamat) [inline src/compiler/corePublic.ts:70]: Is this definition needed  as its already defined above ?

File: src/compiler/moduleNameResolver.ts [hunk 5]
```diff
@@ -472,18 +472,18 @@ namespace ts {
 
     /*@internal*/
     export interface CacheWithRedirects<T> {
-        ownMap: Map<T>;
-        redirectsMap: Map<Map<T>>;
-        getOrCreateMapOfCacheRedirects(redirectedReference: ResolvedProjectReference | undefined): Map<T>;
+        ownMap: Map<string, T>;
+        redirectsMap: Map<string, Map<string, T>>;
```
Comment (sheetalkamat) [inline src/compiler/moduleNameResolver.ts:476]: redirectsMap should be `Map<Path, Map<string, T>>`

File: src/compiler/program.ts [hunk 6]
```diff
@@ -533,7 +533,7 @@ namespace ts {
     export const inferredTypesContainingFile = "__inferred type names__.ts";
 
     interface DiagnosticCache<T extends Diagnostic> {
-        perFile?: Map<readonly T[]>;
+        perFile?: Map<string, readonly T[]>;
```
Comment (sheetalkamat) [inline src/compiler/program.ts:536]: This should be Map<Path, readonly T[]>

File: src/compiler/program.ts [hunk 7]
```diff
@@ -702,14 +702,14 @@ namespace ts {
...
-        let symlinks: ReadonlyMap<string> | undefined;
+        let symlinks: ReadonlyMap<string, string> | undefined;
         let commonSourceDirectory: string;
         let diagnosticsProducingTypeChecker: TypeChecker;
         let noDiagnosticsTypeChecker: TypeChecker;
         let classifiableNames: UnderscoreEscapedMap<true>;
         const ambientModuleNameToUnmodifiedFileName = createMap<string>();
         // Todo:: Use this to report why file was included in --extendedDiagnostics
-        let refFileMap: MultiMap<ts.RefFile> | undefined;
+        let refFileMap: MultiMap<string, ts.RefFile> | undefined;
```
Comment (sheetalkamat) [inline src/compiler/program.ts:712]: This too MultiMap<Path, ts.RefFile>

File: src/compiler/program.ts [hunk 8]
```diff
@@ -803,9 +803,9 @@ namespace ts {
 
         // A parallel array to projectReferences storing the results of reading in the referenced tsconfig files
         let resolvedProjectReferences: readonly (ResolvedProjectReference | undefined)[] | undefined;
-        let projectReferenceRedirects: Map<ResolvedProjectReference | false> | undefined;
-        let mapFromFileToProjectReferenceRedirects: Map<Path> | undefined;
-        let mapFromToProjectReferenceRedirectSource: Map<SourceOfProjectReferenceRedirect> | undefined;
+        let projectReferenceRedirects: Map<string, ResolvedProjectReference | false> | undefined;
```
Comment (sheetalkamat) [inline src/compiler/program.ts:806]: All three shiould have key as Path instead of string

File: src/compiler/program.ts [hunk 9]
```diff
@@ -2892,13 +2892,13 @@ namespace ts {
         function checkSourceFilesBelongToPath(sourceFiles: readonly SourceFile[], rootDirectory: string): boolean {
             let allFilesBelongToPath = true;
             const absoluteRootDirectoryPath = host.getCanonicalFileName(getNormalizedAbsolutePath(rootDirectory, currentDirectory));
-            let rootPaths: Map<true> | undefined;
+            let rootPaths: Set<string> | undefined;
```
Comment (sheetalkamat) [inline src/compiler/program.ts:2895]: Set<Path>

File: src/shims/collectionShims.ts [hunk 10]
```diff
@@ -0,0 +1,443 @@
...
+    function createMapData<K, V>(): MapData<K, V> {
+        const entry: MapEntry<K, V> = {};
+        return { first: entry, last: entry, size: 0 };
+    }
+
+    function createMapEntry<K, V>(key: K, value: V): MapEntry<K, V> {
+        return { key, value, nextIterationEntry: undefined, previousIterationEntry: undefined, skipNextIteration: false };
+    }
+
+    function createListMapEntry<K, V>(key: K, value: V, next?: MapEntry<K, V>): MapEntry<K, V> {
```
Comment (sheetalkamat) [inline src/shims/collectionShims.ts:123]: While allowing different key types seems reasonable, we should do that when we really need it.. We should currently implement only for `T extends string ` keys to keep this simple and compact

Comment (sheetalkamat) [reply src/shims/collectionShims.ts:123]: Infact thats the missing piece.. Previously we ensured we are using string as keys and now we are missing that ability to ensure and make conscious choice of whether we really need map that doesnt key on string

File: src/compiler/binder.ts (hunk 1 above)
Comment (rbuckton) [reply src/compiler/binder.ts:18]: In this case, `visited` isn't a `Map<Path, ...>`. It stores node ids, so it actually should be a `Map<number, ...`>:

https://github.com/microsoft/TypeScript/blob/2f484289c81423bd3939ec7d5431165dc05a111b/src/compiler/binder.ts#L28-L32

Comment (rbuckton) [reply src/compiler/binder.ts:18]: Or, rather, it probably should be a `Map<Node, ...>`, but I plan to look into cases where we use maps with `string`/`number` keys that should actually be `Node` keys at a later date.

Comment (sheetalkamat) [reply src/compiler/binder.ts:18]: Sorry i was looking at builder file and wrote comment at the wrong place.. and later i thought i didnt write it so wrote again..

Comment (sheetalkamat) [reply src/compiler/binder.ts:18]: I meant to write it for builder.ts (For all builder related maps/sets)

File: src/compiler/core.ts (hunk 3 above)
Comment (rbuckton) [reply src/compiler/core.ts:1590]: Not sure. Possibly the line terminator?

File: src/compiler/corePublic.ts (hunk 4 above)
Comment (rbuckton) [reply src/compiler/corePublic.ts:70]: No, its extraneous.

File: src/compiler/program.ts [hunk 11]
```diff
@@ -3204,7 +3204,7 @@ namespace ts {
             // If the emit is enabled make sure that every output file is unique and not overwriting any of the input files
             if (!options.noEmit && !options.suppressOutputPathCheck) {
                 const emitHost = getEmitHost();
-                const emitFilesSeen = createMap<true>();
+                const emitFilesSeen = new Set<string>();
```
[the remark this answers is not in the document]
Comment (rbuckton) [reply src/compiler/program.ts:3207]: This breaks `verifyEmitFilePath` unless we cast the key to `Path`.

File: src/compiler/resolutionCache.ts [hunk 12]
```diff
@@ -161,8 +161,8 @@ namespace ts {
         // The resolvedModuleNames and resolvedTypeReferenceDirectives are the cache of resolutions per file.
         // The key in the map is source file's path.
         // The values are Map of resolutions with key being name lookedup.
-        const resolvedModuleNames = createMap<Map<CachedResolvedModuleWithFailedLookupLocations>>();
-        const perDirectoryResolvedModuleNames: CacheWithRedirects<Map<CachedResolvedModuleWithFailedLookupLocations>> = createCacheWithRedirects();
+        const resolvedModuleNames = createMap<Map<string, CachedResolvedModuleWithFailedLookupLocations>>();
```
[the remark this answers is not in the document]
Comment (rbuckton) [reply src/compiler/resolutionCache.ts:164]: `resolvedModuleNames` currently has the type `Map<string, Map<string, CacheResolved...>>`. Should both keys be `Path`?

File: src/harness/virtualFileSystemWithWatch.ts [hunk 13]
```diff
@@ -122,12 +122,12 @@ interface Array<T> { length: number; [n: number]: T; }`
         }
     }
 
-    function createWatcher<T>(map: MultiMap<T>, path: string, callback: T): FileWatcher {
+    function createWatcher<T>(map: MultiMap<string, T>, path: string, callback: T): FileWatcher {
```
[the remark this answers is not in the document]
Comment (rbuckton) [reply src/harness/virtualFileSystemWithWatch.ts:125]: That breaks the function. I assume `path` should be typed `Path` as well?

File: src/server/editorServices.ts [hunk 14]
```diff
@@ -3075,13 +3075,13 @@ namespace ts.server {
         }
 
         /*@internal*/
-        loadAncestorProjectTree(forProjects?: ReadonlyMap<true>) {
-            forProjects = forProjects || mapDefinedMap(
+        loadAncestorProjectTree(forProjects?: ReadonlyCollection<string>) {
```
[the remark this answers is not in the document]
Comment (rbuckton) [reply src/server/editorServices.ts:3078]: Should `configuredProjects` be a `Map<NormalizedPath, ...>`? If not then this change results in a number of type errors. If so, changing `configuredProjects` to use `NormalizedPath` results in its a number of different type errors.

[the remark this answers is not in the document]
Comment (rbuckton) [reply src/server/editorServices.ts:3078]: To further clarify:
- The type of `configuredProjects` is currently `Map<string, ...>`
- The type of `potentialRefPath` just below this line is `Path`, not `NormalizedPath`.
- The type of `resolvedPath` just below this line is `Path`, not `NormalizedPath`.

File: src/server/project.ts [hunk 15]
```diff
@@ -1814,7 +1814,7 @@ namespace ts.server {
 
         /** Potential project references before the project is actually loaded (read config file) */
         /*@internal*/
-        potentialProjectReferences: Map<true> | undefined;
+        potentialProjectReferences: Set<string> | undefined;
```
[the remark this answers is not in the document]
Comment (rbuckton) [reply src/server/project.ts:1817]: This change breaks `forEachPotentialProjectReference` in editorServices.ts as the callback is typed to accept a `Path`, not a `NormalizedPath`.

File: src/server/editorServices.ts [hunk 16]
```diff
@@ -2296,8 +2296,8 @@ namespace ts.server {
          * Note that this does not return projects in info.containingProjects
          */
         /*@internal*/
-        getSymlinkedProjects(info: ScriptInfo): MultiMap<Project> | undefined {
-            let projects: MultiMap<Project> | undefined;
+        getSymlinkedProjects(info: ScriptInfo): MultiMap<string, Project> | undefined {
```
[the remark this answers is not in the document]
Comment (rbuckton) [reply src/server/editorServices.ts:2299]: This change breaks `callbackProjectAndLocation` in session.ts as the `DocumentPosition` type's `fileName` property is `string`, not `Path`.

File: src/shims/collectionShims.ts (hunk 10 above)
Comment (rbuckton) [reply src/shims/collectionShims.ts:123]: One of the goals of this PR is specifically to allow us to use non-string keys for both `Map` and `Set`, as we have a number of cases where we coerce numbers like symbol IDs, type IDs, and node IDs to string, which is unnecessary overhead. In some cases, we will even be able to use `Map<Node, ...>` or `Set<Node>` instead of allocating a new node ID, and remove the extra cost of those calculations as well.

**Comment by rbuckton** [comment]
@sheetalkamat: I've made most of the suggested key changes, but there are some that would result in type errors that I do not have enough context to resolve.

File: src/server/editorServices.ts (hunk 14 above)
[the remark this answers is not in the document]
Comment (rbuckton) [reply src/server/editorServices.ts:3078]: This also ties into my comment about `DocumentPosition`: https://github.com/microsoft/TypeScript/pull/33771#discussion_r443837400, as the `fileName` property is what feeds the `forProjects` collection that is passed to this method.

**Comment by rbuckton** [comment]
While there are still some outstanding cases to investigate regarding keys, I don't know that they should be a blocker for this PR. Without this PR those keys are currently still typed as `string` anyways, and there is a long tail of cases where we are still using Maps as `Set`-likes, or with string keys that are coerced, that I plan to address over time rather than in a single PR.

**Comment by rbuckton** [comment]
@sheetalkamat can you take another look?

File: src/compiler/program.ts (hunk 11 above)
[the remark this answers is not in the document]
Comment (sheetalkamat) [reply src/compiler/program.ts:3204]: emitFilesSeen.add(emitFileKey); is path in there though.

File: src/harness/virtualFileSystemWithWatch.ts [hunk 17]
```diff
@@ -750,7 +750,7 @@ interface Array<T> { length: number; [n: number]: T; }`
             invokeWatcherCallbacks(this.watchedFiles.get(this.toPath(fileFullPath)), ({ cb, fileName }) => cb(useFileNameInCallback ? fileName : fileFullPath, eventKind));
         }
 
-        private fsWatchCallback(map: MultiMap<TestFsWatcher>, fullPath: string, eventName: "rename" | "change", entryFullPath?: string) {
+        private fsWatchCallback(map: MultiMap<string, TestFsWatcher>, fullPath: string, eventName: "rename" | "change", entryFullPath?: string) {
```
Comment (sheetalkamat) [inline src/harness/virtualFileSystemWithWatch.ts:753]:
```suggestion
        private fsWatchCallback(map: MultiMap<Path, TestFsWatcher>, fullPath: string, eventName: "rename" | "change", entryFullPath?: string) {
```

File: src/compiler/resolutionCache.ts (hunk 12 above)
[the remark this answers is not in the document]
Comment (sheetalkamat) [reply src/compiler/resolutionCache.ts]: Your change is correct because the inner map is for module name

File: src/harness/virtualFileSystemWithWatch.ts (hunk 13 above)
[the remark this answers is not in the document]
Comment (sheetalkamat) [reply src/harness/virtualFileSystemWithWatch.ts]: yes

**Comment by sheetalkamat** [comment]
Sorry didnt notice your questions on the previous comments,going through them now

File: src/server/editorServices.ts (hunk 16 above)
[the remark this answers is not in the document]
Comment (sheetalkamat) [reply src/server/editorServices.ts]: Not sure i understand question here

File: src/server/editorServices.ts (hunk 14 above)
[the remark this answers is not in the document]
Comment (sheetalkamat) [reply src/server/editorServices.ts:3088]: You are right on this one.. Sorry for confusing on this one.

**Comment by rbuckton** [comment]
> Here are two questions i still have:
> 1: There are some places where you replaced createMap() with new Map() but not all. Was there a reason for that

Those were mostly ad-hoc. I am planning on putting together a follow-up PR that replaces all other references and removes `createMap`.

> 2. Do we really care about performance of shim implementaion if we were really using the native maps before.. Why not use .toString() on key for shim to keep it simple and compact if its going to be used rarely.

Using `.toString` won't work for non-primitive values, and wouldn't handle cases like `null`, `undefined`, `true`, `false`, `+Infinity`, `-Infinity`, and `NaN`. The approach in this PR is to give these types different buckets to work with to avoid collisions (i.e. `1` and `"1"` should *not* collide) and still have moderately fast lookup time for common cases (i.e., strings and numbers). 

If the recommendation is to ignore performance for the shim and simplify the design, then I'd likely just do away with the buckets and use the linked list implementation used for `other` values for everything.

One of the stated goals for this PR is to allow us to have non-primitive keys. While I am not making heavy use of it in this PR, I plan to migrate some of our maps/sets to accept `Node` or `Symbol` keys in follow-up PRs, in an effort to reduce the unnecessary overhead of calling `getNodeId` or `getSymbolId` just to use as a key for a Map/Set.

**Comment by rbuckton** [comment]
@sheetalkamat: I've put together a simpler (though less performant) version of the shim that just uses a linked list. Should I use that one instead?

**Comment by rbuckton** [comment]
I'll go ahead and switch to the simpler `Map`/`Set` implementation that just leverages a linked list. It will be less performant, but also takes up less memory and reduces the file-size cost of the shim.

**Comment by rbuckton** [comment]
@sheetalkamat: Would you care to take one more look over the shim implementation? I repurposed the existing doubly-linked list that was added to preserve iteration order, which should result in less memory usage overall.

**Comment by rbuckton** [comment]
@sheetalkamat: I've filed #39264 which leverages the work in this PR and comprehensively removes calls to `createMap`, as well as some other related cleanup.

File: src/shims/collectionShims.ts (hunk 10 above)
Comment (rbuckton) [reply src/shims/collectionShims.ts]: As I've mentioned elsewhere, we already have a `NodeMap` and `NodeSet` class in _utilities.ts_ that has to work around `Map` only supporting `string` keys to act as if it were a native map that allows `Node` as a key. Since we're using a native `Map` under the covers, the custom `NodeMap` and `NodeSet` classes really aren't very useful. #39264 builds on this PR and removes `NodeMap` and `NodeSet` in favor of a regular `Map` and `Set`.
